go/ql/lib/semmle/go/Concepts.qll

```diff
@@ -355,8 +355,19 @@ module RegexpReplaceFunction {
  * extend `LoggerCall::Range` instead.
  */
 class LoggerCall extends DataFlow::Node instanceof LoggerCall::Range {
-  /** Gets a node that is a part of the logged message. */
-  DataFlow::Node getAMessageComponent() { result = super.getAMessageComponent() }
+  /**
+   * Gets a node that is a part of the logged message.
+   *
+   * Exclude components corresponding to the format specifier "%T" as this
+   * prints the type of the argument, which is not considered vulnerable.
+   */
+  DataFlow::Node getAMessageComponent() {
+    result = super.getAMessageComponent() and
+    not exists(string formatSpecifier |
+      formatSpecifier.regexpMatch("%[^%]*T") and
+      result = this.(StringOps::Formatting::StringFormatCall).getOperand(_, formatSpecifier)
+    )
+  }
```

Comment:
Is this the right place for this? Given that it's in `Concepts`, I'd think this should be more general than what we need for one specific query?

Reply:
This predicate is used in three places (outside of tests). Two of them are sinks of logging queries, where we want to exclude %T arguments. The last one is trying to rule out oauth flows that use printer calls and scan calls, which makes it likely that it's done in a terminal, and I think for that purpose it will make no difference but it's probably slightly better to rule out %T arguments. It seemed good to catch at least the first two cases. The alternative would be to add another predicate, `getAPrintedMessageComponent`, which has this new restriction, but the result would just be that we don't use `getAMessageComponent` anywhere any more. So I think it's okay to change the meaning of this predicate since it's making it closer to what anyone would want to use it for.

Reply:
My concern is that introducing query-specific logic here could be quite confusing for anyone who wants to use `getAMessageComponent` in the future to query all the components of a `LoggerCall`. It's easy to miss the comment that explains that it has non-intuitive behaviour and it might not be obvious during testing (of a new query). So while it might be fine for current uses, I don't think it's a good move for maintainability.

Rather than adding a new predicate to `LoggerCall`, I think it would be better to add a new class that extends `LoggerCall`s which is specific to the `go/clear-text-logging` query (and others) where we then override `getAMessageComponent` and filter `%T` components out.

Reply:
I don't agree. No one cares about things which have their type logged. I can make a new class, but it means the old one will literally not be used in our library. Anyone who wants that behaviour can make a copy of `LoggerCall` and change the behaviour of `getAMessageComponent`. I will make the comment more prominent so that the intention is clearer.